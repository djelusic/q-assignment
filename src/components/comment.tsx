import { Comment } from "../state/main";
import { withLogging, WithLoggingProps } from "../util/withLogging";
import '../css/comment.scss'

interface CommentProps extends WithLoggingProps {
  comment: Comment
}

export const CommentComponent = withLogging(({ comment }: CommentProps) => {
  return (
    <article className="comment">
      <div className="avatar-container">
        <span className="avatar">
          {comment.name[0].toUpperCase()}
        </span>
      </div>
      <div className="body-container">
        <h4 className="name">{comment.name}</h4>
        <p className="body">{comment.body}</p>
      </div>
    </article>
  )
})
CommentComponent.displayName = 'Comment'
