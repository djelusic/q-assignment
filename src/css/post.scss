@import "./avatar.scss";

.post {
  background-color: white;
  border-radius: 24px;
  padding: 12px;
  .user {
    display: flex;
    align-items: center;
    .avatar {
      margin-right: 8px;
    }
    .username {
      font-size: 13px;
      font-weight: bold;
    }
  }
  .title {
    margin: 12px 0;
  }
  .body {
    font-size: 14px;
  }
  .comments-button {
    margin-top: 8px;
    font-size: 12px;
    color: #8ab4f8;
    text-align: right;
  }
  .comments {
    .comment {
      margin-bottom: 12px;
    }
  }
}
