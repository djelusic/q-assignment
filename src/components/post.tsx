import { useCallback, useState } from 'react'

import { Post } from '../state/main'
import { withLogging, WithLoggingProps } from '../util/withLogging'
import { CommentComponent } from './comment'
import '../css/post.scss'

interface PostProps extends WithLoggingProps {
  post: Post
  alwaysShowComments?: boolean
  onPostClick?: () => void
}

export const PostComponent = withLogging(({ post, alwaysShowComments, onPostClick, loggingPrefix }: PostProps) => {
  const [showComments, setShowComments] = useState(false)
  const toggleShowComments = useCallback(() => {
    setShowComments(!showComments);
  }, [showComments])

  const hasComments = post.comments?.length ?? 0 > 0
  return (
    <article className="post">
      <div onClick={() => onPostClick?.()}>
        <div className="user">
          <span className="avatar">
            {post.username[0].toUpperCase()}
          </span>
          <span className="username">{post.username}</span>
        </div>
        <h4 className="title">{post.title}</h4>
        <p className="body">{post.body}</p>
      </div>
      { hasComments && !alwaysShowComments ?
        <div className="comments-button" onClick={toggleShowComments}>
          {showComments ? 'Hide comments' : `Show comments (${post.comments?.length})`}
        </div>
        : null}
      { hasComments && (showComments || alwaysShowComments) ?
        <div className="comments">
          <h4>Comments</h4>
          {post.comments?.map(comment => {
            return (
              <CommentComponent
                key={comment.id}
                comment={comment}
                loggingPrefix={loggingPrefix}
              />
            )
          })}
        </div> : null
      }
    </article>
  )
})
PostComponent.displayName = 'Post'
