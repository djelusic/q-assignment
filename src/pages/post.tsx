import { useCallback } from "react";
import { useParams } from "react-router";

import { IState, Post } from "../state/main";
import { PostComponent } from "../components/post"
import { Loader } from "../util/loader"
import { withLogging, WithLoggingProps } from "../util/withLogging";

interface PostPageComponentProps extends WithLoggingProps {
  data: Post
}

const PostPageComponent = withLogging(({ data, loggingPrefix }: PostPageComponentProps) => {
  return (
    <>
      <h2>Post</h2>
      <PostComponent
        post={data}
        alwaysShowComments
        loggingPrefix={loggingPrefix}
      />
    </>
  )
})
PostPageComponent.displayName = 'PostPageComponent'

interface PostPageProps extends WithLoggingProps {
  state: IState
}

export const PostPage = withLogging(({ state, loggingPrefix }: PostPageProps) => {
  const { id } = useParams()
  const fetcher = useCallback(() => {
    return state.fetchPost(parseInt(id ?? '0'))
  }, [id, state.fetchPost])
  return (
    <Loader
      fetcher={fetcher}
      render={(data => <PostPageComponent data={data} loggingPrefix={loggingPrefix} />)}
    />
  )
})
PostPage.displayName = 'PostPage'
