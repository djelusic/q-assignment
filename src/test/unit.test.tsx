import { fireEvent, render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { PostComponent } from '../components/post'

test('post', async () => {
  const { asFragment } = render(
    <PostComponent post={{
      id: 1,
      userId: 1,
      username: 'User',
      title: 'Title',
      body: 'Body',
      comments: [
        {
          postId: 1,
          id: 1,
          name: 'Name',
          body: 'Body',
        },
      ],
    }} loggingPrefix="Test" />
  )
  expect(asFragment()).toMatchSnapshot()
  
  expect(screen.getByText('Show comments (1)')).toBeInTheDocument()
  expect(screen.queryByText('Hide comments')).toBeNull()
  expect(screen.queryByText('Comments')).toBeNull()

  fireEvent.click(screen.getByText('Show comments (1)'))
  expect(screen.queryByText('Show comments (1)')).toBeNull()
  expect(screen.getByText('Hide comments')).toBeInTheDocument()
  expect(screen.getByText('Comments')).toBeInTheDocument()
})
