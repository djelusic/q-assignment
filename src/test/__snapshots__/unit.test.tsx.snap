// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post 1`] = `
<DocumentFragment>
  <article
    class="post"
  >
    <div>
      <div
        class="user"
      >
        <span
          class="avatar"
        >
          U
        </span>
        <span
          class="username"
        >
          User
        </span>
      </div>
      <h4
        class="title"
      >
        Title
      </h4>
      <p
        class="body"
      >
        Body
      </p>
    </div>
    <div
      class="comments-button"
    >
      Show comments (1)
    </div>
  </article>
</DocumentFragment>
`;
